### PR TITLE
Make AzureKeyVaultConfigurationProvider public

### LINKED ISSUE #15459: [FEATURE REQ] AzureKeyVaultConfigurationSource should not create a new config provider during Build
In `Azure.Extensions.AspNetCore.Configuration.Secrets`, a new `AzureKeyVaultConfigurationProvider` is [created in the `AzureKeyVaultConfigurationSource.Build` method](https://github.com/Azure/azure-sdk-for-net/blob/a3b6a5ff0f73177571aeaf5670f5fbe753d942fb/sdk/extensions/Azure.Extensions.AspNetCore.Configuration.Secrets/src/AzureKeyVaultConfigurationSource.cs#L21):

```csharp
public IConfigurationProvider Build(IConfigurationBuilder builder)
{
    return new AzureKeyVaultConfigurationProvider(_options.Client, _options.Manager, _options.ReloadInterval);
}
```

The problem with this is that, while configuring my ASP.NET Core app, I need to build the ConfigurationBuilder multiple times. This is because our KeyVault settings are set in environment variables + appsettings.json, and our Azure App Configuration connection string is stored in KeyVault.

So basically I need to:
```csharp
Host.CreateDefaultBuilder()
    .ConfigureAppConfiguration(configBuilder =>
    {
        var config = configBuilder.Build(); // load KeyVault settings from default config locations
        var keyVaultOpt = config.GetSection("KeyVault").Get<KeyVaultOptions>();

        var cred = new ClientSecretCredential(keyVaultOpt.TenantId, keyVaultOpt.ApplicationId, keyVaultOpt.Secret);
        configBuilder.AddAzureKeyVault(new Uri(keyVaultOpt.Url), cred);

        config = configBuilder.Build(); // load AppConfig settings from KeyVault
        var appConfigOpt = config.GetSection("AppConfig").Get<AppConfigOptions>();

        configBuilder.AddAzureAppConfiguration(appConfigOpt.ConnectionString);
    });
```

However since `AzureKeyVaultConfigurationSource` creates a new `AzureKeyVaultConfigurationProvider`, two separate providers will be created: once during my second `configBuilder.Build()`, and once when the app builds it for the final time. That means that my KeyVault secrets are loaded twice, causing unnecessary delays during startup.

If instead, `AzureKeyVaultConfigurationSource` used a `Lazy<T>` and only created it once:
```csharp
internal class AzureKeyVaultConfigurationSource : IConfigurationSource
{
    private readonly AzureKeyVaultConfigurationOptions _options;
    private readonly Lazy<AzureKeyVaultConfigurationProvider> _providerLazy;

    public AzureKeyVaultConfigurationSource(AzureKeyVaultConfigurationOptions options)
    {
        _options = options;
        _providerLazy = new Lazy<AzureKeyVaultConfigurationProvider>(() =>
            new AzureKeyVaultConfigurationProvider(_options.Client, _options.Manager, _options.ReloadInterval));
    }

    /// <inheritdoc />
    public IConfigurationProvider Build(IConfigurationBuilder builder)
    {
        return _providerLazy.Value;
    }
}
```

Then the provider wouldn't be re-created and (as far as I can tell) the secrets are loaded only once.

I'm open to other suggestions on how to better accomplish this; I realize that building the `ConfigurationBuilder` multiple times isn't the standard practice, but I feel like storing the App Config connection string in Key Vault has to be a somewhat common practice so maybe there's a better way?

It would also be very helpful if `AzureKeyVaultConfigurationProvider` (and `AzureKeyVaultConfigurationSource` for that matter) were `public` instead of `internal` so I could accomplish this within my app, if there is a reason this request can't implemented.

### LINKED ISSUE #19638: [FEATURE REQ] add `DataChanged` method to `KeyVaultSecretManager` to support custom load logic
**Library or service name.**
What library or service is this request related to? Azure.Extensions.AspNetCore.Configuration.Secrets
 
**Is your feature request related to a problem? Please describe.**
`KeyVaultSecretManager` has `GetKey` and `Load` methods to customize the behavior of the `AzureKeyVaultConfigurationProvider` but there are other scenarios such as #11943 which require more complex custom logic.

Rather than have the provider try to support all scenarios, end-users could implement their own custom logic if `KeyVaultSecretManager` included a method such as `public virtual void DataChanged(Dictionary<string, LoadedSecret> loadedSecrets)` which defaults to a no-op in the default implementation.

This could be called [immediately after loading secrets](https://github.com/Azure/azure-sdk-for-net/blob/bb7d612114a8a6c8e5ce8c55a9359cded69488dd/sdk/extensions/Azure.Extensions.AspNetCore.Configuration.Secrets/src/AzureKeyVaultConfigurationProvider.cs#L107) but immediately before setting the `Data` and issuing reload notices. This will allow custom logic to modify/add/remove dictionary entries as needed, before they're actually set on the provider.

```csharp
foreach (var secretBundle in loadedSecret)
{
    newLoadedSecrets.Add(secretBundle.Value.Name, new LoadedSecret(_manager.GetKey(secretBundle), secretBundle.Value.Value, secretBundle.Value.Properties.UpdatedOn));
}

_manager.DataChanged(newLoadedSecrets); // allow KeyVaultSecretManager to execute custom logic on all loaded secrets

_loadedSecrets = newLoadedSecrets;
```

### PR DESCRIPTION
**Pull request by pakrym**
We had a few request from people wrapping the AzureKeyVaultConfigurationProvider and needing it to be public and others who wanted advanced transformation of secrets they've received (parse single JSON value into multiple configuration keys, etc.).

This PRs does both. 

Fixes: https://github.com/Azure/azure-sdk-for-net/issues/19638
Fixes: https://github.com/Azure/azure-sdk-for-net/issues/15459

### CODE REVIEW

File: sdk/extensions/Azure.Extensions.AspNetCore.Configuration.Secrets/src/AzureKeyVaultConfigurationProvider.cs [hunk 1]
```diff
@@ -15,36 +15,36 @@ namespace Azure.Extensions.AspNetCore.Configuration.Secrets
...
-        /// <param name="manager">The <see cref="KeyVaultSecretManager"/> to use in managing values.</param>
-        /// <param name="reloadInterval">The timespan to wait in between each attempt at polling the Azure Key Vault for changes. Default is null which indicates no reloading.</param>
-        public AzureKeyVaultConfigurationProvider(SecretClient client, KeyVaultSecretManager manager, TimeSpan? reloadInterval = null)
+        /// <param name="options">The <see cref="AzureKeyVaultConfigurationOptions"/> to configure provider behaviors.</param>
+        public AzureKeyVaultConfigurationProvider(SecretClient client, AzureKeyVaultConfigurationOptions options = null)
         {
+            options ??= new AzureKeyVaultConfigurationOptions();
             Argument.AssertNotNull(client, nameof(client));
-            Argument.AssertNotNull(manager, nameof(manager));
+            Argument.AssertNotNull(options.Manager, $"{nameof(options)}.{nameof(options.Manager)}");
```
Comment (heaths) [inline sdk/extensions/Azure.Extensions.AspNetCore.Configuration.Secrets/src/AzureKeyVaultConfigurationProvider.cs:36]: Exceptions should be documented, especially now that it's public.

Nit: *technically* this shouldn't be an `ArgumentNullException` and instead an `InvalidOperationException`, but I doubt there's any significant difference for what's primarily a dev-time error (i.e. should never happen in production).

File: sdk/extensions/Azure.Extensions.AspNetCore.Configuration.Secrets/src/AzureKeyVaultConfigurationProvider.cs [hunk 2]
```diff
@@ -15,36 +15,36 @@ namespace Azure.Extensions.AspNetCore.Configuration.Secrets
...
 
             _client = client;
-            _manager = manager;
-            if (reloadInterval != null && reloadInterval.Value <= TimeSpan.Zero)
+            if (options.ReloadInterval != null && options.ReloadInterval.Value <= TimeSpan.Zero)
             {
-                throw new ArgumentOutOfRangeException(nameof(reloadInterval), reloadInterval, nameof(reloadInterval) + " must be positive.");
+                throw new ArgumentOutOfRangeException(nameof(options.ReloadInterval), options.ReloadInterval, nameof(options.ReloadInterval) + " must be positive.");
```
Comment (heaths) [inline sdk/extensions/Azure.Extensions.AspNetCore.Configuration.Secrets/src/AzureKeyVaultConfigurationProvider.cs:41]: Nit: same as above.

File: sdk/extensions/Azure.Extensions.AspNetCore.Configuration.Secrets/src/AzureKeyVaultConfigurationProvider.cs [hunk 3]
```diff
@@ -125,68 +129,22 @@ await foreach (var secret in secretPages.ConfigureAwait(false))
...
-            Data = data.ToDictionary(d => d.Key, v => v.Value.Value);
-            if (fireToken)
-            {
-                OnReload();
-            }
-        }
-
         /// <inheritdoc/>
-        public void Dispose()
+        void IDisposable.Dispose()
```
Comment (heaths) [inline sdk/extensions/Azure.Extensions.AspNetCore.Configuration.Secrets/src/AzureKeyVaultConfigurationProvider.cs:133]: Are people meant to be able to derive this class now? If so, this should probably use the protected `Disposing(bool)` pattern.

File: sdk/extensions/Azure.Extensions.AspNetCore.Configuration.Secrets/src/KeyVaultSecretManager.cs [hunk 4]
```diff
@@ -24,6 +27,39 @@ public virtual string GetKey(KeyVaultSecret secret)
             return secret.Name.Replace("--", ConfigurationPath.KeyDelimiter);
         }
 
+        /// <summary>
+        /// Converts a loaded list of secrets into a corresponding set of configuration key-value pairs.
+        /// </summary>
+        /// <param name="secrets">A set of secrets retrieved during <see cref="AzureKeyVaultConfigurationProvider.Load"/> call.</param>
+        /// <returns>The dictionary of configuration key-value pairs that would be assigned to the <see cref="ConfigurationProvider.Data"/>
+        /// and exposed from the <see cref="IConfiguration"/>.</returns>
+        public virtual Dictionary<string, string> GetData(IEnumerable<KeyVaultSecret> secrets)
```
Comment (heaths) [inline sdk/extensions/Azure.Extensions.AspNetCore.Configuration.Secrets/src/KeyVaultSecretManager.cs:36]: Public API should check arguments. Be sure to document them as well.

**Review by heaths — APPROVED** [review]
A few nits, but nothing I'd hold the PR for.

**Comment by JohnGalt1717** [comment]
Thank you so much! We were blocked switching because of this!

**Pull request merged**